dh/sidh/internal/templates/curve_test.gotemp

```diff
@@ -98,3 +102,207 @@ func BenchmarkThreePointLadder(b *testing.B) {
 		ScalarMul3Pt(&curve, &threePointLadderInputs[0], &threePointLadderInputs[1], &threePointLadderInputs[2], uint(len(scalar3Pt)*8), scalar3Pt[:])
 	}
 }
+
+/* -------------------------------------------------------------------------
+   Generate invalid public key points / ciphertext for test TestKEMInvalidPK
+   -------------------------------------------------------------------------*/
+
+// Given the curve parameters, xP = x(P), computes xP = x([2]P)
+// Safe to overlap xP, x2P.
+func xDbl(xP *ProjectivePoint, params *CurveCoefficientsEquiv) {
+	var t0, t1 Fp2
+
+	x, z := &xP.X, &xP.Z
+	sub(&t0, x, z)           // t0  = Xp - Zp
+	add(&t1, x, z)           // t1  = Xp + Zp
+	sqr(&t0, &t0)            // t0  = t0 ^ 2
+	sqr(&t1, &t1)            // t1  = t1 ^ 2
+	mul(z, &params.C, &t0)   // Z2p = C24 * t0
+	mul(x, z, &t1)           // X2p = Z2p * t1
+	sub(&t1, &t1, &t0)       // t1  = t1 - t0
+	mul(&t0, &params.A, &t1) // t0  = A24+ * t1
+	add(z, z, &t0)           // Z2p = Z2p + t0
+	mul(z, z, &t1)           // Zp  = Z2p * t1
+}
+
+
+
+// Right-to-left Montgomery ladder, Algorithm 4 in Costello-Smith
+// Input: ProjectivePoint P (xP, zP)
+// Output: x([scalar]P), z([scalar]P)
+func montgomeryLadder(cparams *ProjectiveCurveParameters, P *ProjectivePoint, scalar []uint8, random uint) ProjectivePoint {
+	var R0, R2, R1 ProjectivePoint
+    coefEq := CalcCurveParamsEquiv4(cparams) // for xDbl
+	aPlus2Over4 := CalcAplus2Over4(cparams)	 // for xDblAdd
+	R0 = *P					 // RO <- P
+	R1 = *P; xDbl(&R1, &coefEq)		 // R1 <- [2]P
+	R2 = *P					 // R2 = R1-R0 = P
+
+	prevBit := uint8(0)
+	for i := int(random); i >= 0; i-- {
+		bit := (scalar[i>>3] >> (i & 7) & 1)
+		swap := prevBit ^ bit
+		prevBit = bit
+		cswap(&R0.X, &R0.Z, &R1.X, &R1.Z, swap)
+		R0, R1 = xDbladd(&R0, &R1, &R2, &aPlus2Over4)
+	}
+	cswap(&R0.X, &R0.Z, &R1.X, &R1.Z, prevBit)
+	return R0
+}
+
+// P = P + T
+func tauT(P *ProjectivePoint) {
+	P.X, P.Z = P.Z, P.X	// magic!
+}
+
+// Construct Invalid public key tuple (P,Q) such that P and Q are linearly dependent 
+// Simulate section 3.1.1 of paper https://eprint.iacr.org/2022/054.pdf
+// We only construct point P and Q because in the attacks the third point is P-Q by construction 
+// and the countermeasure does not test it
+// Without loss of generality, we assume the curve is the starting curve 
+func InvalidPKNoneLinear(scalar []uint8, pub3Pt *[3]Fp2) {
+	
+	var P, Q ProjectivePoint
+	
+	rand.Seed(time.Now().UnixNano())
```

Comment:
we may want to keep track of the seed that was used, this is in the case that the test fails so we can reproduce it.